File: Source/AlphaTab.JavaScript/Platform/JavaScript/Html5Canvas.cs
using System;
using AlphaTab.Haxe.Js;
using AlphaTab.Haxe.Js.Html;
using AlphaTab.Platform.Model;
using AlphaTab.Rendering.Glyphs;
using TextAlign = AlphaTab.Platform.Model.TextAlign;

namespace AlphaTab.Platform.JavaScript
{
    /// <summary>
    /// A canvas implementation for HTML5 canvas
    /// </summary>
    internal class Html5Canvas : ICanvas
    {
        protected const float BlurCorrection = 0;

        private CanvasElement _measureCanvas;
        private CanvasRenderingContext2D _measureContext;

        private CanvasElement _canvas;
        private CanvasRenderingContext2D _context;
        private Color _color;
        private Font _font;
        private Font _musicFont;
        private float _lineWidth;

        public Settings Settings { get; set; }

        public Html5Canvas()
        {
            _color = new Color(0, 0, 0);
            var fontElement = Browser.Document.CreateElement("span");
            fontElement.ClassList.Add("at");
            Browser.Document.Body.AppendChild(fontElement);
            var style = Browser.Window.GetComputedStyle(fontElement);
            string family = style.FontFamily;
            if (family.StartsWith("\"") || family.StartsWith("'"))
            {
                family = family.Substring(1, family.Length - 2);
            }

            _musicFont = new Font(family, Platform.ParseFloat(style.FontSize));

            _measureCanvas = (CanvasElement)Browser.Document.CreateElement("canvas");
            _measureCanvas.Width = 10;
            _measureCanvas.Height = 10;
            _measureCanvas.Style.Width = 10 + "px";
            _measureCanvas.Style.Height = 10 + "px";
            _measureContext = (CanvasRenderingContext2D)_measureCanvas.GetContext("2d");
            _measureContext.TextBaseline = "top";
        }

        public virtual object OnRenderFinished()
        {
            // nothing to do
            return null;
        }

        public void BeginRender(float width, float height)
        {
            _canvas = (CanvasElement)Browser.Document.CreateElement("canvas");
            _canvas.Width = (int)width;
            _canvas.Height = (int)height;
            _canvas.Style.Width = width + "px";
            _canvas.Style.Height = height + "px";
            _context = (CanvasRenderingContext2D)_canvas.GetContext("2d");
            _context.TextBaseline = "top";
            _context.LineWidth = _lineWidth;
        }

        public object EndRender()
        {
            var result = _canvas;
            _canvas = null;
            return result;
        }

        public Color Color
        {
            get => _color;
            set
            {
                if (_color.Rgba == value.Rgba)
                {
                    return;
                }

                _color = value;
                _context.StrokeStyle = value.Rgba;
                _context.FillStyle = value.Rgba;
            }
        }

        public float LineWidth
        {
            get => _lineWidth;
            set
            {
                _lineWidth = value;
                if (_context != null)
                {
                    _context.LineWidth = value;
                }
            }
        }

        public void FillRect(float x, float y, float w, float h)
        {
            if (w > 0)
            {
                _context.FillRect(((int)x - BlurCorrection), ((int)y - BlurCorrection), w, h);
            }
        }

        public void StrokeRect(float x, float y, float w, float h)
        {
            _context.StrokeRect((x - BlurCorrection), (y - BlurCorrection), w, h);
        }

        public void BeginPath()
        {
            _context.BeginPath();
        }

        public void ClosePath()
        {
            _context.ClosePath();
        }

        public void MoveTo(float x, float y)
        {
            _context.MoveTo((x - BlurCorrection), (y - BlurCorrection));
        }

        public void LineTo(float x, float y)
        {
            _context.LineTo((x - BlurCorrection), (y - BlurCorrection));
        }

        public void QuadraticCurveTo(float cpx, float cpy, float x, float y)
        {
            _context.QuadraticCurveTo(cpx, cpy, x, y);
        }

        public void BezierCurveTo(float cp1x, float cp1y, float cp2x, float cp2y, float x, float y)
        {
            _context.BezierCurveTo(cp1x, cp1y, cp2x, cp2y, x, y);
        }

        public void FillCircle(float x, float y, float radius)
        {
            _context.BeginPath();
            _context.Arc(x, y, radius, 0, (Math.PI * 2), true);
            Fill();
        }

        public void Fill()
        {
            _context.Fill();
        }

        public void Stroke()
        {
            _context.Stroke();
        }

        public Font Font
        {
            get => _font;
            set
            {
                _font = value;
                if (_context != null)
                {
                    _context.Font = value.ToCssString(Settings.Display.Scale);
                }

                _measureContext.Font = value.ToCssString(Settings.Display.Scale);
            }
        }

        public TextAlign TextAlign
        {
            get
            {
                switch (_context.TextAlign)
                {
                    case "left":
                        return TextAlign.Left;
                    case "center":
                        return TextAlign.Center;
                    case "right":
                        return TextAlign.Right;
                    default:
                        return TextAlign.Left;
                }
            }
            set
            {
                switch (value)
                {
                    case TextAlign.Left:
                        _context.TextAlign = "left";
                        break;
                    case TextAlign.Center:
                        _context.TextAlign = "center";
                        break;
                    case TextAlign.Right:
                        _context.TextAlign = "right";
                        break;
                }
            }
        }

        public TextBaseline TextBaseline
        {
            get
            {
                switch (_context.TextBaseline)
                {
                    case "top":
                        return TextBaseline.Top;
                    case "middle":
                        return TextBaseline.Middle;
                    case "bottom":
                        return TextBaseline.Bottom;
                    default:
                        return TextBaseline.Top;
                }
            }
            set
            {
                switch (value)
                {
                    case TextBaseline.Top:
                        _context.TextBaseline = "top";
                        break;
                    case TextBaseline.Middle:
                        _context.TextBaseline = "middle";
                        break;
                    case TextBaseline.Bottom:
                        _context.TextBaseline = "bottom";
                        break;
                }
            }
        }

        public void BeginGroup(string identifier)
        {
        }

        public void EndGroup()
        {
        }

        public void FillText(string text, float x, float y)
        {
            x = (int)x;
            y = (int)y;
            _context.FillText(text, x, y);
        }

        public float MeasureText(string text)
        {
            return (float)_measureContext.MeasureText(text).Width;
        }

        public void FillMusicFontSymbol(
            float x,
            float y,
            float scale,
            MusicFontSymbol symbol,
            bool centerAtPosition = false)
        {
            if (symbol == MusicFontSymbol.None)
            {
                return;
            }

            FillMusicFontSymbolText(x, y, scale, Platform.StringFromCharCode((int)symbol), centerAtPosition);
        }

        public void FillMusicFontSymbols(
            float x,
            float y,
            float scale,
            MusicFontSymbol[] symbols,
            bool centerAtPosition = false)
        {
            var s = "";
            foreach (var symbol in symbols)
            {
                if (symbol != MusicFontSymbol.None)
                {
                    s += Platform.StringFromCharCode((int)symbol);
                }
            }

            FillMusicFontSymbolText(x, y, scale, s, centerAtPosition);
        }

        private void FillMusicFontSymbolText(
            float x,
            float y,
            float scale,
            string symbols,
            bool centerAtPosition = false)
        {
            x = (int)x;
            y = (int)y;
            var textAlign = _context.TextAlign;
            var baseLine = _context.TextBaseline;
            var font = _context.Font;
            _context.Font = _musicFont.ToCssString(scale);
            _context.TextBaseline = "middle";
            if (centerAtPosition)
            {
                _context.TextAlign = "center";
            }

            _context.FillText(symbols, x, y);
            _context.TextBaseline = baseLine;
            _context.Font = font;
            _context.TextAlign = textAlign;
        }

        public void BeginRotate(float centerX, float centerY, float angle)
        {
            _context.Save();
            _context.Translate(centerX, centerY);
            _context.Rotate(angle * Math.PI / 180.0f);
        }

        public void EndRotate()
        {
            _context.Restore();
        }
    }
}

File: src/platform/javascript/Html5Canvas.ts
import { Environment } from '@src/Environment';
import { Color } from '@src/model/Color';
import { Font, FontStyle } from '@src/model/Font';
import { MusicFontSymbol } from '@src/model/MusicFontSymbol';
import { type ICanvas, TextAlign, TextBaseline, MeasuredText } from '@src/platform/ICanvas';
import type { Settings } from '@src/Settings';

/**
 * A canvas implementation for HTML5 canvas
 * @target web
 */
export class Html5Canvas implements ICanvas {
    private _measureCanvas: HTMLCanvasElement;
    private _measureContext: CanvasRenderingContext2D;
    private _canvas: HTMLCanvasElement | null = null;
    private _context!: CanvasRenderingContext2D;
    private _color: Color = new Color(0, 0, 0, 0xff);
    private _font: Font = new Font('Arial', 10, FontStyle.Plain);
    private _musicFont?: Font;
    private _lineWidth: number = 0;

    public settings!: Settings;

    public constructor() {
        this._measureCanvas = document.createElement('canvas');
        this._measureCanvas.width = 10;
        this._measureCanvas.height = 10;
        this._measureCanvas.style.width = '10px';
        this._measureCanvas.style.height = '10px';
        this._measureContext = this._measureCanvas.getContext('2d')!;
        this._measureContext.textBaseline = 'hanging';
    }

    public destroy() {}

    public onRenderFinished(): unknown {
        return null;
    }

    public beginRender(width: number, height: number): void {
        this._musicFont = this.settings.display.resources.smuflFont;

        const scale = this.settings.display.scale;

        this._canvas = document.createElement('canvas');
        this._canvas.width = (width * Environment.HighDpiFactor) | 0;
        this._canvas.height = (height * Environment.HighDpiFactor) | 0;
        this._canvas.style.width = `${width}px`;
        this._canvas.style.height = `${height}px`;
        this._context = this._canvas.getContext('2d')!;
        this._context.textBaseline = 'hanging';
        this._context.scale(Environment.HighDpiFactor * scale, Environment.HighDpiFactor * scale);
        this._context.lineWidth = this._lineWidth;
    }

    public endRender(): unknown {
        const result: HTMLCanvasElement = this._canvas!;
        this._canvas = null;
        return result;
    }

    public get color(): Color {
        return this._color;
    }

    public set color(value: Color) {
        if (this._color.rgba === value.rgba) {
            return;
        }
        this._color = value;
        this._context.strokeStyle = value.rgba;
        this._context.fillStyle = value.rgba;
    }

    public get lineWidth(): number {
        return this._lineWidth;
    }

    public set lineWidth(value: number) {
        this._lineWidth = value;
        if (this._context) {
            this._context.lineWidth = value;
        }
    }

    public fillRect(x: number, y: number, w: number, h: number): void {
        if (w > 0) {
            this._context.fillRect(x | 0, y | 0, w, h);
        }
    }

    public strokeRect(x: number, y: number, w: number, h: number): void {
        const blurOffset = this.lineWidth % 2 === 0 ? 0 : 0.5;
        this._context.strokeRect((x | 0) + blurOffset, (y | 0) + blurOffset, w, h);
    }

    public beginPath(): void {
        this._context.beginPath();
    }

    public closePath(): void {
        this._context.closePath();
    }

    public moveTo(x: number, y: number): void {
        this._context.moveTo(x, y);
    }

    public lineTo(x: number, y: number): void {
        this._context.lineTo(x, y);
    }

    public quadraticCurveTo(cpx: number, cpy: number, x: number, y: number): void {
        this._context.quadraticCurveTo(cpx, cpy, x, y);
    }

    public bezierCurveTo(cp1x: number, cp1y: number, cp2x: number, cp2y: number, x: number, y: number): void {
        this._context.bezierCurveTo(cp1x, cp1y, cp2x, cp2y, x, y);
    }

    public fillCircle(x: number, y: number, radius: number): void {
        this._context.beginPath();
        this._context.arc(x, y, radius, 0, Math.PI * 2, true);
        this.fill();
    }

    public strokeCircle(x: number, y: number, radius: number): void {
        this._context.beginPath();
        this._context.arc(x, y, radius, 0, Math.PI * 2, true);
        this.stroke();
    }

    public fill(): void {
        this._context.fill();
        this._context.beginPath();
    }

    public stroke(): void {
        this._context.stroke();
        this._context.beginPath();
    }

    public get font(): Font {
        return this._font;
    }

    public set font(value: Font) {
        this._font = value;
        if (this._context) {
            this._context.font = value.toCssString(1);
        }
        this._measureContext.font = value.toCssString(1);
    }

    public get textAlign(): TextAlign {
        switch (this._context.textAlign) {
            case 'left':
                return TextAlign.Left;
            case 'center':
                return TextAlign.Center;
            case 'right':
                return TextAlign.Right;
            default:
                return TextAlign.Left;
        }
    }

    public set textAlign(value: TextAlign) {
        switch (value) {
            case TextAlign.Left:
                this._context.textAlign = 'left';
                break;
            case TextAlign.Center:
                this._context.textAlign = 'center';
                break;
            case TextAlign.Right:
                this._context.textAlign = 'right';
                break;
        }
    }

    public get textBaseline(): TextBaseline {
        switch (this._context.textBaseline) {
            case 'hanging':
                return TextBaseline.Top;
            case 'middle':
                return TextBaseline.Middle;
            case 'bottom':
                return TextBaseline.Bottom;
            default:
                return TextBaseline.Top;
        }
    }

    public set textBaseline(value: TextBaseline) {
        switch (value) {
            case TextBaseline.Top:
                this._context.textBaseline = 'hanging';
                break;
            case TextBaseline.Middle:
                this._context.textBaseline = 'middle';
                break;
            case TextBaseline.Bottom:
                this._context.textBaseline = 'bottom';
                break;
        }
    }

    public beginGroup(_: string): void {
        // not supported
    }

    public endGroup(): void {
        // not supported
    }

    public fillText(text: string, x: number, y: number): void {
        this._context.fillText(text, x, y);
    }

    public measureText(text: string): MeasuredText {
        const metrics = this._measureContext.measureText(text);
        return new MeasuredText(metrics.width, metrics.actualBoundingBoxDescent + metrics.actualBoundingBoxAscent);
    }

    public fillMusicFontSymbol(
        x: number,
        y: number,
        relativeScale: number,
        symbol: MusicFontSymbol,
        centerAtPosition: boolean = false
    ): void {
        if (symbol === MusicFontSymbol.None) {
            return;
        }
        this.fillMusicFontSymbolText(x, y, relativeScale, String.fromCharCode(symbol), centerAtPosition);
    }

    public fillMusicFontSymbols(
        x: number,
        y: number,
        relativeScale: number,
        symbols: MusicFontSymbol[],
        centerAtPosition: boolean = false
    ): void {
        let s: string = '';
        for (const symbol of symbols) {
            if (symbol !== MusicFontSymbol.None) {
                s += String.fromCharCode(symbol);
            }
        }
        this.fillMusicFontSymbolText(x, y, relativeScale, s, centerAtPosition);
    }

    private fillMusicFontSymbolText(
        x: number,
        y: number,
        relativeScale: number,
        symbols: string,
        centerAtPosition: boolean
    ): void {
        const textAlign = this._context.textAlign;
        const baseLine = this._context.textBaseline;
        const font: string = this._context.font;
        this._context.font = this._musicFont!.toCssString(relativeScale);
        this._context.textBaseline = 'middle';
        if (centerAtPosition) {
            this._context.textAlign = 'center';
        } else {
            this._context.textAlign = 'left';
        }
        this._context.fillText(symbols, x, y);
        this._context.textBaseline = baseLine;
        this._context.font = font;
        this._context.textAlign = textAlign;
    }

    public beginRotate(centerX: number, centerY: number, angle: number): void {
        this._context.save();
        this._context.translate(centerX, centerY);
        this._context.rotate((angle * Math.PI) / 180.0);
    }

    public endRotate(): void {
        this._context.restore();
    }
}
